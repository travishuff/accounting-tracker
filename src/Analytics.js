import React from 'react'
import axios from 'axios'
import {
  differenceInCalendarDays,
  format,
  lastDayOfMonth,
  startOfMonth,
} from 'date-fns'
import get from 'lodash/get'
import { Link } from '@reach/router'

import './css/analytics.css'

import { Button, Container, Form, Table } from 'semantic-ui-react'
import Dates from './Dates'
import Margins from './Margins'

class Analytics extends React.Component {
  state = {
    data: [],
    buyPrice: 0.2,
    sellPrice: 0.35,
    start: format(startOfMonth(new Date()), 'YYYY-MM-DD'),
    end: format(lastDayOfMonth(new Date()), 'YYYY-MM-DD'),
  }

  bananasByTime = () => {
    return this.state.data.filter(banana => {
      const buyDate = get(banana, 'buyDate', '')
      const sellDate = get(banana, 'sellDate', '')

      return (
        Date.parse(buyDate) >= Date.parse(this.state.start) &&
        Date.parse(buyDate) <= Date.parse(this.state.end) &&
        (sellDate === null ||
          (Date.parse(sellDate) >= Date.parse(this.state.start) &&
            Date.parse(sellDate) <= Date.parse(this.state.end)))
      )
    })
  }
  expiredBananas = () => {
    return this.bananasByTime().filter(banana => {
      return differenceInCalendarDays(new Date(), get(banana, 'buyDate')) >= 11
    })
  }
  soldBananas = () => {
    return this.bananasByTime().filter(
      banana => get(banana, 'sellDate') !== null
    )
  }
  unexpiredBananas = () => {
    return this.bananasByTime().filter(banana => {
      return differenceInCalendarDays(new Date(), get(banana, 'buyDate')) < 11
    })
  }
  unsoldBananas = () => {
    return this.bananasByTime().filter(
      banana => get(banana, 'sellDate') === null
    )
  }

  // GAAP Measures: computed properties based on current state.
  soldBananasValue = () => {
    return (this.soldBananas().length * this.state.sellPrice).toFixed(2)
  }
  totalBananasCost = () => {
    return (this.bananasByTime().length * this.state.buyPrice).toFixed(2)
  }
  totalProfit = () => {
    return (this.soldBananasValue() - this.totalBananasCost()).toFixed(2)
  }

  // Non-GAAP Measures: computed properties based on current state.
  unsoldUnexpiredBananas = () => {
    return this.unexpiredBananas().filter(banana => banana.sellDate === null)
  }
  unsoldUnexpiredBananasValue = () => {
    // prettier-ignore
    return (this.unsoldUnexpiredBananas().length * this.state.sellPrice).toFixed(2)
  }
  unsoldExpiredBananas = () => {
    return this.expiredBananas().filter(banana => banana.sellDate === null)
  }
  unsoldExpiredBananasCost = () => {
    return (this.unsoldExpiredBananas().length * this.state.buyPrice).toFixed(2)
  }
  allOtherBananasCost = () => {
    // prettier-ignore
    return (this.totalBananasCost() - this.unsoldExpiredBananasCost()).toFixed(2)
  }
  potentialProfit = () => {
    return (
      Number(this.soldBananasValue()) +
      Number(this.unsoldUnexpiredBananasValue()) -
      Number(this.totalBananasCost())
    ).toFixed(2)
  }

  handlePriceChange = event => {
    this.setState({
      [event.target.name]: Number(event.target.value),
    })

    localStorage.setItem(event.target.name, event.target.value)
  }

  handleDateChange = event => {
    this.setState({
      [event.target.name]: event.target.value,
    })

    localStorage.setItem(event.target.name, event.target.value)
  }

  resetFields = () => {
    this.setState({
      buyPrice: 0.2,
      sellPrice: 0.35,
      start: format(startOfMonth(new Date()), 'YYYY-MM-DD'),
      end: format(lastDayOfMonth(new Date()), 'YYYY-MM-DD'),
    })

    localStorage.clear()
  }

  hydrateStateWithLocalStorage() {
    for (let key in this.state) {
      if (localStorage.hasOwnProperty(key)) {
        let value = localStorage.getItem(key)

        try {
          value = JSON.parse(value)
          this.setState({ [key]: value })
        } catch (e) {
          this.setState({ [key]: value })
        }
      }
    }
  }

  async componentDidMount() {
    const response = await axios
      .get('http://localhost:8080/api/bananas')
      .catch(console.error)
    this.setState(
      {
        data: response?.data,
      },
      this.hydrateStateWithLocalStorage()
    )
  }

  render() {
    const { buyPrice, sellPrice, start, end } = this.state

    return (
      <Container fluid={true} className="main-container">
        <h1>Analytics</h1>

        <Dates
          start={start}
          end={end}
          handleDateChange={this.handleDateChange}
        />

        <Table color="blue" unstackable>
          <Table.Header>
            <Table.Row>
              <Table.HeaderCell colSpan="4">Non-GAAP Measures</Table.HeaderCell>
            </Table.Row>
            <Table.Row>
              <Table.HeaderCell>Item</Table.HeaderCell>
              <Table.HeaderCell>Amount</Table.HeaderCell>
              <Table.HeaderCell>Price</Table.HeaderCell>
              <Table.HeaderCell>Value</Table.HeaderCell>
            </Table.Row>
          </Table.Header>
          <Table.Body>
            <Table.Row>
              <Table.Cell>Bananas sold</Table.Cell>
              <Table.Cell>{this.soldBananas().length}</Table.Cell>
              <Table.Cell>${sellPrice.toFixed(2)}</Table.Cell>
              <Table.Cell positive>${this.soldBananasValue()}</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Unsold unexpired bananas</Table.Cell>
              <Table.Cell>{this.unsoldUnexpiredBananas().length}</Table.Cell>
              <Table.Cell>${sellPrice.toFixed(2)}</Table.Cell>
              <Table.Cell positive>
                ${this.unsoldUnexpiredBananasValue()}
              </Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Unsold expired bananas</Table.Cell>
              <Table.Cell>{this.unsoldExpiredBananas().length}</Table.Cell>
              <Table.Cell>${buyPrice.toFixed(2)}</Table.Cell>
              <Table.Cell negative>
                ${this.unsoldExpiredBananasCost()}
              </Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>All other purchased bananas</Table.Cell>
              <Table.Cell>
                {this.bananasByTime().length -
                  this.unsoldExpiredBananas().length}
              </Table.Cell>
              <Table.Cell>${buyPrice.toFixed(2)}</Table.Cell>
              <Table.Cell negative>${this.allOtherBananasCost()}</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell colSpan="3" textAlign="right">
                Potential Profit/Loss
              </Table.Cell>
              <Table.Cell
                positive={this.potentialProfit() > 0 ? true : false}
                negative={this.potentialProfit() < 0 ? true : false}
              >
                ${Math.abs(this.potentialProfit()).toFixed(2)}
              </Table.Cell>
            </Table.Row>
          </Table.Body>
        </Table>

        <Table color="green" unstackable>
          <Table.Header>
            <Table.Row>
              <Table.HeaderCell colSpan="4">GAAP Measures</Table.HeaderCell>
            </Table.Row>
            <Table.Row>
              <Table.HeaderCell>Item</Table.HeaderCell>
              <Table.HeaderCell>Amount</Table.HeaderCell>
              <Table.HeaderCell>Price</Table.HeaderCell>
              <Table.HeaderCell>Value</Table.HeaderCell>
            </Table.Row>
          </Table.Header>
          <Table.Body>
            <Table.Row>
              <Table.Cell>Bananas sold</Table.Cell>
              <Table.Cell>{this.soldBananas().length}</Table.Cell>
              <Table.Cell>${sellPrice.toFixed(2)}</Table.Cell>
              <Table.Cell positive>${this.soldBananasValue()}</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell>Bananas purchased</Table.Cell>
              <Table.Cell>{this.bananasByTime().length}</Table.Cell>
              <Table.Cell>${buyPrice.toFixed(2)}</Table.Cell>
              <Table.Cell negative>${this.totalBananasCost()}</Table.Cell>
            </Table.Row>
            <Table.Row>
              <Table.Cell colSpan="3" textAlign="right">
                Profit/Loss
              </Table.Cell>
              <Table.Cell
                positive={this.totalProfit() > 0 ? true : false}
                negative={this.totalProfit() < 0 ? true : false}
              >
                ${Math.abs(this.totalProfit()).toFixed(2)}
              </Table.Cell>
            </Table.Row>
          </Table.Body>
        </Table>

        <Button color="blue">
          <Link to="/groups" className="button-text">
            List of Banana Groups
          </Link>
        </Button>
        <Button color="blue">
          <Link to="/list" className="button-text">
            Full List of Bananas
          </Link>
        </Button>

        <Margins
          buyPrice={buyPrice}
          sellPrice={sellPrice}
          handlePriceChange={this.handlePriceChange}
        />

        <Form.Button
          color="red"
          onClick={this.resetFields}
          content="Reset All Form Fields"
        />
      </Container>
    )
  }
}

export default Analytics
