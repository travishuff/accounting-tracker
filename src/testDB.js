export const DB = [
  {
    id: 'ba7e45a1-2bf0-4050-a331-ca67d29b033e',
    buyDate: '2019-05-24',
    sellDate: '2019-05-25',
  },
  {
    id: '9708b11f-6295-4980-bddf-75d4b73ec461',
    buyDate: '2019-05-24',
    sellDate: '2019-05-26',
  },
  {
    id: '7b7731ad-e4ad-41d8-b417-5caa3dc44b89',
    buyDate: '2019-05-25',
    sellDate: null,
  },
  {
    id: '93b03cdc-1480-4220-861c-e20756a43e42',
    buyDate: '2019-06-01',
    sellDate: '2019-06-06',
  },
  {
    id: 'eb95cd5b-d56c-4dd8-86bd-73ba09c9272e',
    buyDate: '2019-06-30',
    sellDate: '2019-07-02',
  },
  {
    id: '04f42e82-dd56-4e4f-851a-9cb29eb27bd7',
    buyDate: '2019-06-30',
    sellDate: '2019-07-01',
  },
  {
    id: '19d1adf5-cb4d-42f3-b8b2-116f435c0523',
    buyDate: '2019-06-30',
    sellDate: null,
  },
  {
    id: '7d18746c-8db0-46a3-8160-6033beb2c4bb',
    buyDate: '2019-07-01',
    sellDate: null,
  },
]
