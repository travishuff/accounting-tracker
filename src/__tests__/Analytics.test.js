import React from 'react'
import { create } from 'react-test-renderer'

import Analytics from '../Analytics'
import { DB } from '../testDB'

// Snapshot test
test('snapshot', () => {
  const c = create(<Analytics />)

  expect(c.toJSON()).toMatchSnapshot()
})

// Unit testing
test('handlePriceChange for `buyPrice` method changes value', () => {
  const c = create(<Analytics />)
  const instance = c.getInstance()

  instance.state.buyPrice = 0.25
  const event = {
    target: {
      name: 'buyPrice',
      value: 0.5,
    },
  }
  instance.handlePriceChange(event)
  expect(instance.state.buyPrice).toBe(0.5)
})

test('handlePriceChange for `sellPrice` method changes value', () => {
  const c = create(<Analytics />)
  const instance = c.getInstance()

  instance.state.sellPrice = 0.35
  const event = {
    target: {
      name: 'sellPrice',
      value: 0.5,
    },
  }
  instance.handlePriceChange(event)
  expect(instance.state.sellPrice).toBe(0.5)
})

test('handleDateChange for `start` method changes value', () => {
  const c = create(<Analytics />)
  const instance = c.getInstance()

  instance.state.start = '2019-06-01'
  const event = {
    target: {
      name: 'start',
      value: '2019-06-30',
    },
  }
  instance.handleDateChange(event)
  expect(instance.state.start).toBe('2019-06-30')
})

test('handleDateChange for `end` method changes value', () => {
  const c = create(<Analytics />)
  const instance = c.getInstance()

  instance.state.end = '2019-06-01'
  const event = {
    target: {
      name: 'end',
      value: '2019-06-30',
    },
  }
  instance.handleDateChange(event)
  expect(instance.state.end).toBe('2019-06-30')
})

test('bananasByTime method creates array', () => {
  const c = create(<Analytics />)
  const instance = c.getInstance()

  instance.state.start = '2019-06-01'
  instance.state.end = '2019-06-30'
  instance.state.data = DB
  const result = instance.bananasByTime()
  expect(result).toStrictEqual([
    {
      id: '93b03cdc-1480-4220-861c-e20756a43e42',
      buyDate: '2019-06-01',
      sellDate: '2019-06-06',
    },
    {
      id: '19d1adf5-cb4d-42f3-b8b2-116f435c0523',
      buyDate: '2019-06-30',
      sellDate: null,
    },
  ])
})
