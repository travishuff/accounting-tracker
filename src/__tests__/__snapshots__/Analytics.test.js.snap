// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  className="ui fluid container main-container"
>
  <h1>
    Analytics
  </h1>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="fields"
    >
      <div
        className="four wide field"
      >
        <label>
          Start Date:
        </label>
        <div
          className="ui input"
        >
          <input
            name="start"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
            value="2019-05-01"
          />
        </div>
      </div>
      <div
        className="four wide field"
      >
        <label>
          End Date:
        </label>
        <div
          className="ui input"
        >
          <input
            name="end"
            onBlur={[Function]}
            onChange={[Function]}
            type="date"
            value="2019-05-31"
          />
        </div>
      </div>
    </div>
  </form>
  <table
    className="ui blue unstackable table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className=""
          colSpan="4"
        >
          Non-GAAP Measures
        </th>
      </tr>
      <tr
        className=""
      >
        <th
          className=""
        >
          Item
        </th>
        <th
          className=""
        >
          Amount
        </th>
        <th
          className=""
        >
          Price
        </th>
        <th
          className=""
        >
          Value
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    >
      <tr
        className=""
      >
        <td
          className=""
        >
          Bananas sold
        </td>
        <td
          className=""
        >
          0
        </td>
        <td
          className=""
        >
          $
          0.35
        </td>
        <td
          className="positive"
        >
          $
          0.00
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
        >
          Unsold unexpired bananas
        </td>
        <td
          className=""
        >
          0
        </td>
        <td
          className=""
        >
          $
          0.35
        </td>
        <td
          className="positive"
        >
          $
          0.00
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
        >
          Unsold expired bananas
        </td>
        <td
          className=""
        >
          0
        </td>
        <td
          className=""
        >
          $
          0.20
        </td>
        <td
          className="negative"
        >
          $
          0.00
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
        >
          All other purchased bananas
        </td>
        <td
          className=""
        >
          0
        </td>
        <td
          className=""
        >
          $
          0.20
        </td>
        <td
          className="negative"
        >
          $
          0.00
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className="right aligned"
          colSpan="3"
        >
          Potential Profit/Loss
        </td>
        <td
          className=""
        >
          $
          0.00
        </td>
      </tr>
    </tbody>
  </table>
  <table
    className="ui green unstackable table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className=""
          colSpan="4"
        >
          GAAP Measures
        </th>
      </tr>
      <tr
        className=""
      >
        <th
          className=""
        >
          Item
        </th>
        <th
          className=""
        >
          Amount
        </th>
        <th
          className=""
        >
          Price
        </th>
        <th
          className=""
        >
          Value
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    >
      <tr
        className=""
      >
        <td
          className=""
        >
          Bananas sold
        </td>
        <td
          className=""
        >
          0
        </td>
        <td
          className=""
        >
          $
          0.35
        </td>
        <td
          className="positive"
        >
          $
          0.00
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className=""
        >
          Bananas purchased
        </td>
        <td
          className=""
        >
          0
        </td>
        <td
          className=""
        >
          $
          0.20
        </td>
        <td
          className="negative"
        >
          $
          0.00
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className="right aligned"
          colSpan="3"
        >
          Profit/Loss
        </td>
        <td
          className=""
        >
          $
          0.00
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="ui blue button"
    onClick={[Function]}
  >
    <a
      className="button-text"
      href="/groups"
      onClick={[Function]}
    >
      List of Banana Groups
    </a>
  </button>
  <button
    className="ui blue button"
    onClick={[Function]}
  >
    <a
      className="button-text"
      href="/list"
      onClick={[Function]}
    >
      Full List of Bananas
    </a>
  </button>
  <h3>
    Experiment with your margins!
  </h3>
  <form
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="fields"
    >
      <div
        className="four wide field"
      >
        <label>
          Buy Price:
        </label>
        <div
          className="ui input"
        >
          <input
            name="buyPrice"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value={0.2}
          />
        </div>
      </div>
      <div
        className="four wide field"
      >
        <label>
          Sell Price:
        </label>
        <div
          className="ui input"
        >
          <input
            name="sellPrice"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value={0.35}
          />
        </div>
      </div>
    </div>
  </form>
  <div
    className="field"
  >
    <button
      className="ui red button"
      onClick={[Function]}
    >
      Reset All Form Fields
    </button>
  </div>
</div>
`;
